R/hard_group_by.r

```diff
@@ -108,10 +109,18 @@ hard_group_by.data.frame <- function(df, ..., add = FALSE, .drop = FALSE) {
 #' @importFrom purrr map
 #' @importFrom purrr map
 #' @export
-hard_group_by.disk.frame <- function(df, ..., outdir=tempfile("tmp_disk_frame_hard_group_by"), nchunks = disk.frame::nchunks(df), overwrite = TRUE) {
+hard_group_by.disk.frame <- function(df, ..., outdir=tempfile("tmp_disk_frame_hard_group_by"), nchunks = disk.frame::nchunks(df), overwrite = TRUE, shardby_function="hash", sort_splits=NULL, desc_vars=NULL) {
   overwrite_check(outdir, overwrite)
   
   ff = list.files(attr(df, "path"))
+  stopifnot(shardby_function %in% c("hash", "sort"))
+  
+  if (shardby_function == "sort" && is.null(sort_splits)){
+    sort_splits <- map(df, sample_n, size=ceiling(nchunks / disk.frame::nchunks(df))) %>% 
```

Comment:
can we sample more? I think sampling `nchunks` rows from each `chunk`, collect them and then pick out the cut points from the `nchunks^2` rows would be better.

Also, what if there are `duplicated` from each sample? I think it will still work, but we may want to generate some warnings in some cases.

Reply:
Added more sampling - default 100 per. Open to suggestions on how to generate warnings!